ssh/tunnel.py

```diff
@@ -81,62 +44,76 @@ def remote_cmd(self, cmd, timeout=None, stdout=None):
             logging.debug('Timed out process output:\n' + e.output.decode())
             raise
 
-    def write_to_remote(self, src, dst):
+    def write_to_remote(self, src: str, dst: str):
         """
         Args:
-            src: (str) local path representing source data
-            dst: (str) destination for path
+            src: local path representing source data
+            dst: destination for path
         """
-        cmd = self.cmd + ['-C', '-p', str(self.port), self.target, 'cat>' + dst]
+        cmd = self.base_cmd + ['-C', self.target, 'cat>' + dst]
         logger.debug('Running socket write: ' + ' '.join(cmd))
         with open(src, 'r') as fh:
             check_call(cmd, stdin=fh)
 
-    def close(self):
-        close_tunnel = self.cmd + ['-p', str(self.port), '-O', 'exit', self.target]
-        logger.debug('Closing SSH Tunnel: ' + ' '.join(close_tunnel))
-        check_call(close_tunnel)
 
+@contextmanager
+def tunnel(user, key_path, host, port=22):
+    tunnel_socket = tempfile.NamedTemporaryFile()
+    target = user + '@' + host
+
+    base_cmd = [
+        '/usr/bin/ssh',
+        '-oConnectTimeout=10',
+        '-oControlMaster=auto',
+        '-oControlPath=' + tunnel_socket.name,
+        '-oStrictHostKeyChecking=no',
+        '-oUserKnownHostsFile=/dev/null',
+        '-oBatchMode=yes',
+        '-oPasswordAuthentication=no',
+        '-p', str(port)]
+
+    start_tunnel = base_cmd + ['-fnN', '-i', key_path, target]
+    logger.debug('Starting SSH tunnel: ' + ' '.join(start_tunnel))
+    check_call(start_tunnel)
+    logger.debug('SSH Tunnel established!')
+
+    yield Tunnelled(base_cmd, host, target)
+
+    close_tunnel = base_cmd + ['-O', 'exit', target]
+    logger.debug('Closing SSH Tunnel: ' + ' '.join(close_tunnel))
+    check_call(close_tunnel)
+
+
+@contextmanager
+def tunnel_collection(user: str, key_path: str, host_names: list):
+    """Convenience collection of Tunnels so that users can keep
+    multiple connections alive with a single self-closing context
+    Args:
+        user: user with access to host
+        key_path: local path w/ permissions to user@host
+        host_names: list of locally resolvable hostname:port to tunnel to
+    """
 
-class TunnelCollection():
```

Comment:
:+1:  I really like the conversion of an extremely specific class to a simple context manager method